File: leetcode/0078.subsets/javascript/001.backtracking.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var subsets = function(nums) {
  const subsets = [];

  const subset = [];
  const backtrack = (idx) => {
    if(idx >= nums.length) {
      subsets.push([...subset]);
      return;
    }

    subset.push(nums[idx]);
    backtrack(idx+1);

    subset.pop();
    backtrack(idx+1);
  }

  backtrack(0);

  return subsets;
};

// Time: O(n*2^n)
// Space: O(n)

File: leetcode/0078.subsets/rust/001.backtracking.rs
impl Solution {
    pub fn subsets(nums: Vec<i32>) -> Vec<Vec<i32>> {
        let mut results = Vec::new();
        let mut subset = Vec::new();

        fn dfs(
          i: usize,
          nums: &[i32],
          subset: &mut Vec<i32>,
          results: &mut Vec<Vec<i32>>, 
        ) {
            if i >= nums.len() {
                results.push(subset.clone());
                return
            }

            subset.push(nums[i]);
            dfs(i+1, nums, subset, results);
            
            subset.pop();
            dfs(i+1, nums, subset, results);
        }


        dfs(0, &nums, &mut subset, &mut results);

        results
    }
}
